Xamarin.Forms.Core/Shell/ShellUriHandler.cs

```diff
@@ -404,26 +476,39 @@ internal static List<RouteRequestBuilder> GenerateRoutePaths(Shell shell, Uri re
 		internal static List<string> CollapsePath(
 				string myRoute,
 				IEnumerable<string> currentRouteStack,
-				bool userDefinedRoute)
+				bool userDefinedRoute) =>
+			CollapsePath(RetrievePaths(myRoute), currentRouteStack, userDefinedRoute);
+
+
+		internal static List<string> CollapsePath(
+				string[] myRoute,
+				IEnumerable<string> currentRouteStack,
+				bool removeUserDefinedRoute)
 		{
 			var localRouteStack = currentRouteStack.ToList();
 			for (var i = localRouteStack.Count - 1; i >= 0; i--)
 			{
 				var route = localRouteStack[i];
 				if (Routing.IsImplicit(route) ||
-					(Routing.IsDefault(route) && userDefinedRoute))
+					(Routing.IsDefault(route) && removeUserDefinedRoute))
 				{
 					localRouteStack.RemoveAt(i);
 				}
 			}
 
-			var paths = RetrievePaths(myRoute).ToList();
+			var paths = myRoute.ToList();
 
 			// collapse similar leaves
-			int walkBackCurrentStackIndex = localRouteStack.Count - (paths.Count - 1);
+			int walkBackCurrentStackIndex = -1;
+
+			if (paths.Count > 0)
+				walkBackCurrentStackIndex = localRouteStack.IndexOf(paths[0]);
```

Comment:
This is new code

The idea here is to collapse the URI

If a user is at

`//monkey/page/nextpage`

And they have a global route that's

`page/nextpage/something`

Then this code will combine those two paths so they become

`//monkey/page/something`